homeassistant/components/mqtt/__init__.py

```diff
@@ -246,8 +250,14 @@ def async_mqtt_topic_subscriber(dp_topic, dp_payload, dp_qos):
     async_remove = async_dispatcher_connect(
         hass, SIGNAL_MQTT_MESSAGE_RECEIVED, async_mqtt_topic_subscriber)
 
+    @callback
+    def async_remove_unsubscribe():
+        """Unsubscribe when removing signal listener."""
+        hass.async_run_job(hass.data[DATA_MQTT].async_unsubscribe(topic, qos))
```

Comment:
It's messy that now we have 2 places where subscriptions live, that feels weird.

Probably out of scope, but it would almost make more sense to get rid of the dispatcher and move topic matching and calling listeners into our MQTT client. 

Something like this:

```python
sub_info = namedtuple('SubscriptionInfo', 'callback,qos,topic,encoding')

for sub in self._subscriptions.items():
    if not _match_topic(sub.topic, topic):
        return
    …
    self.hass.async_run_job(sub.callback, topic, payload, qos)
```

Reply:
Yes, it's probably a good time to clean up the subscription model a bit. 🎨